In [7]:
import pandas as pd
from mod06_deanonymize import load_data, link_records, deanonymization_rate

In [8]:
anon_path = "mod06_data/anonymized.csv"
aux_path = "mod06_data/auxiliary.csv"

### Load and review the data

In [9]:
anon, aux = load_data(anon_path, aux_path)

In [10]:
anon.head(20)

,anon_id,age,zip3,gender
0,ERM5C,50,432,F
1,OFJJP,75,323,F
2,I3H24,44,323,M
3,3WUYW,66,323,M
4,ZN91P,18,323,M
5,7BLKV,36,606,F
6,4SJFH,70,890,M
7,0NO3L,31,110,M
8,U060Z,80,191,F
9,1761H,58,606,M


### Merge the datasets by age, gender, and zip3 and return the unique matches

In [11]:
matches = link_records(anon, aux)
matches.head()

,anon_id,matched_name
0,ERM5C,Grace Lopez
13,ZN91P,Kyle Martinez
32,0NO3L,Marcus Lopez
33,U060Z,Sally Davis
34,1761H,Frank Jackson


### Calculate the match rate

In [12]:
rate = deanonymization_rate(matches, anon)
print(rate)

0.421


# Discussion Questions

### Describe your reaction to the match rate. What sources of data might be prone to privacy issues?

The match rate of 0.421 shows that a large portion of the anonymized records could still be re-identified using only a few demographic variables. This demonstrates that removing names alone does not fully protect privacy because combinations of attributes such as age, gender, and ZIP code can uniquely identify individuals. Many real-world datasets could be vulnerable to this type of re-identification, especially datasets that contain demographic or location information.

### Provide examples of at least two other variables that could increase the match rate if included.

Additional variables would likely increase the match rate because they make each record more unique. For example, including full ZIP code instead of ZIP3 would make geographic identification more precise. Another variable that could increase the match rate is date of birth, which is often highly identifying when combined with other demographic attributes.